src/com/ms/silverking/cloud/zookeeper/ZooKeeperExtended.java

```diff
@@ -42,1024 +43,1081 @@
-                    } finally {
-                        lock.unlock();
-                    }
-                    if (displayMissingSW.getSplitSeconds() > displayMissingIntervalSeconds) {
-                        displayMissing = true;
-                        displayMissingSW.reset();
-                    }
-                } catch (InterruptedException ie) {
-                }
-                if (!complete) {
-                    if (getState() != ZooKeeper.States.CONNECTED) {
-                        throw KeeperException.create(Code.SESSIONEXPIRED);
-                    }
-                }
-                if (!complete) {
-                    checkForCompletion(displayMissing);
-                }
-            }
-        }
 
-        private void checkForCompletion(boolean displayMissing) {
-            if (observer == null || observer.isActive()) {
-                int        numIncomplete;
-                
-                numIncomplete = 0;
-                for (String path : paths) {
-                    if (!dataMap.containsKey(path) && !errorKeys.contains(path)) {
-                        ++numIncomplete;
-                        if (displayMissing) {
-                            Log.warning("checkForCompletion still waiting for: ", path);
-                        }
-                        break;
-                    }
-                }
-                if (numIncomplete == 0) {
-                    complete = true;
-                } else {
-                    if (retrySW.getSplitSeconds() > reissueIntervalSeconds) {
-                        if (timeoutSW.getSplitSeconds() > timeoutSeconds) {
-                            markIncompleteAsErrors();
-                        } else {
-                            issueGetsForIncomplete();
-                        }
-                    }
-                }
-            }
-        }
-        
-        /**
-         * Notify waiter. Let the waiter check for completion, resend etc.
-         * We want to keep callback operations as light as possible.
-         */
-        private void notifyWaiter() {
-            lock.lock();
-            try {
-                cv.signalAll();
-            } finally {
-                lock.unlock();
-            }
-            
+    // precondition: only a single caller allowed in this method
+    public void waitForCompletion() throws KeeperException {
+      while (!complete && (observer == null || observer.isActive())) {
+        boolean displayMissing;
+
+        displayMissing = false;
+        try {
+          lock.lock();
+          try {
+            cv.await(displayMissingIntervalSeconds, TimeUnit.SECONDS);
+          } finally {
+            lock.unlock();
+          }
+          if (displayMissingSW.getSplitSeconds() > displayMissingIntervalSeconds) {
+            displayMissing = true;
+            displayMissingSW.reset();
+          }
+        } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)